### PR TITLE
Improvements to factor parsing

### PR DESCRIPTION
**Pull request by jimhester**
This includes the changes in https://github.com/tidyverse/readr/pull/602, https://github.com/tidyverse/readr/pull/592 as well as the fix for https://github.com/tidyverse/readr/issues/615.

### CODE REVIEW

File: tests/testthat/test-parsing-factors.R [hunk 1]
```diff
@@ -17,8 +17,39 @@ test_that("warning if value not in levels", {
...
+
+  expect_equal(n_problems(x), 0)
+  expect_equal(x, factor(c("a", "b", "c", "b")))
+})
+
+test_that("NAs included in levels if desired", {
+  x <- parse_factor(c("NA", "b", "a"), levels = c("a", "b", NA), include_na = TRUE)
+  expect_equal(x, factor(c("NA", "b", "a"), exclude = NULL))
+
+  expect_warning(x <- parse_factor(c("NA", "b", "a"), levels = c("a", "b"), include_na = TRUE))
```
Comment (hadley) [inline tests/testthat/test-parsing-factors.R]: I don't think this should generate a warning.

Comment (jimhester) [inline tests/testthat/test-parsing-factors.R]: Why not? you are saying you want to include NA in the levels, but don't actually provide a NA level...

Comment (hadley) [inline tests/testthat/test-parsing-factors.R]: I think NA is a special case. But I might be overthinking it

Comment (jimhester) [inline tests/testthat/test-parsing-factors.R]: Ok thinking about this more I think you are right, `include_na = TRUE` now will include the NA level if there is an NA in the input and it is not in the levels.

**Pull request merged**